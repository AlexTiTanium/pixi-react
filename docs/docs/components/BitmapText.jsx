import { BitmapText, Stage, Text } from '@pixi/react';
import ExampleAssetLoader from './ExampleAssetLoader';

const position = {
    anchor: 0.5,
    x: 250,
    y: 150,
};
export default function App() {
    return (
        <Stage width={500} height={300} options={{ backgroundColor: 0xffffff }}>
            <ExampleAssetLoader
                name='desyrel'
                url='https://pixijs.io/pixi-react/font/desyrel.xml'
                loader={<Text {...position} text='âŒ› Loading font...' style={{ fontFamily: 'Arial', fontSize: 15 }} />}
            >
                <BitmapText {...position} text='Hello World!' style={{ fontName: 'Desyrel', fontSize: 50 }} />
            </ExampleAssetLoader>
        </Stage>
    );
}
