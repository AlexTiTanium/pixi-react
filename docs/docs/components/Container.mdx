import { EmbeddedEditor } from '@site/src/components/Editor/EmbeddedEditor';
import IndexFile from '!!raw-loader!./Container';

# Container

## Props

https://pixijs.download/v7.x/docs/PIXI.Container.html

## Usage

<EmbeddedEditor height={350} files={{ 'App.js': IndexFile }} />

## Example

<iframe
  height={500}
  scrolling="no"
  title="Container"
  src="//codepen.io/inlet/embed/ca683993e355b9f3370f0a44d189141f/?height=300&theme-id=33987&default-tab=result&embed-version=2"
  frameBorder="no"
  allowFullScreen={true}
  style={{ width: '100%' }}
/>
