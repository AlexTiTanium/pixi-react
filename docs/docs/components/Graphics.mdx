import { EmbeddedEditor } from '@site/src/components/Editor/EmbeddedEditor';
import IndexFile from '!!raw-loader!./Graphics';

# Graphics

https://pixijs.download/v7.x/docs/PIXI.Graphics.html

## Props

| prop     | description            |
| -------- | ---------------------- |
| draw     | Draw callback function |
| geometry | Graphics object        |

## Usage

<EmbeddedEditor height={350} files={{ 'App.js': IndexFile }} />

### The `draw` prop

Note that the `draw` callback gets called everytime the prop changes, make sure to memoize the callback
with `React.useCallback`:

```jsx
import React, { useCallback } from 'react';
import { Graphics } from '@pixi/react';

function Rectangle(props) {
  const draw = useCallback(
    (g) => {
      g.clear();
      g.beginFill(props.color);
      g.drawRect(props.x, props.y, props.width, props.height);
      g.endFill();
    },
    [props],
  );

  return <Graphics draw={draw} />;
}
```

### The `geometry` prop

Provides another Graphics object as a template. Helps in reducing memory footprint if the same shapes are used repeatedly.

```jsx
import React, { useCallback } from 'react';
import { Graphics } from '@pixi/react';

function Rectangle(props) {
  const draw = useCallback((g) => {
    g.clear();
    g.lineStyle(props.lineWidth, props.color);
    g.drawRect(
      props.lineWidth,
      props.lineWidth,
      props.width - 2 * props.lineWidth,
      props.height - 2 * props.lineWidth
    );
  }, [props]);

  return <Graphics draw={draw} />
}

function Grid(props) {
  const rect = useRef();
  const [mounted, setMounted] = React.useState(false);

  React.useEffect(() => {
    setMounted(true);
  }, []);

  return (
    <>
      {/* create the graphics component here */}
      <Rectangle ref={rect}>

      {/* make sure to wait till `Rectangle` is mounted */}
      {/* geometry can only be set during component creation */}
      {mounted && (
        <>
          <Graphics x={0} y={0} geometry={rect.current} />
          <Graphics x={props.width} y={0} geometry={rect.current} />
          <Graphics x={0} y={props.height} geometry={rect.current} />
          <Graphics x={props.width} y={props.height} geometry={rect.current} />
        </>
      )}
    </>
  );
}
```

## Example

<iframe
  height={500}
  scrolling="no"
  title="Graphics"
  src="//codepen.io/inlet/embed/d3b1be3b908b1d890442ee0c8baceaaf/?height=300&theme-id=33987&default-tab=result&embed-version=2"
  frameBorder="no"
  allowFullScreen={true}
  style={{ width: '100%' }}
/>
