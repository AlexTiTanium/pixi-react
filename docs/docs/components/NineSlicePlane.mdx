import { EmbeddedEditor } from '@site/src/components/Editor/EmbeddedEditor';
import IndexFile from '!!raw-loader!./NineSlicePlane';

# NineSlicePlane

## Props

https://pixijs.download/v7.x/docs/PIXI.NineSlicePlane.html

## Usage

<EmbeddedEditor height={350} files={{ 'App.js': IndexFile }} />

## Example

<iframe
  height={500}
  scrolling="no"
  title="Nine Slice Plane"
  src="//codepen.io/inlet/embed/8bf47b3e9a4539c472ea2836780c59eb/?height=300&theme-id=33987&default-tab=result&embed-version=2"
  frameBorder="no"
  allowFullScreen={true}
  style={{ width: '100%' }}
/>
