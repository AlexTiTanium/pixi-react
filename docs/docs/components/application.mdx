---
sidebar_position: 1
title: Application
---

import { EmbeddedEditor } from '@site/src/components/Editor/EmbeddedEditor';
import ExampleAppBasic from '!!raw-loader!../../src/examples/application-basic/App';
import ExampleAppResizeTo from '!!raw-loader!../../src/examples/application-resizeTo/App';

The `<Application>` component is used to wrap your `@pixi/react` app. The `<Application>` component can take [all props that can be set][pixi-ApplicationOptions] on [`PIXI.Application`][pixi-Application].

## Example Usage

<EmbeddedEditor
    files={{
        'App.js': ExampleAppBasic,
    }}
    viewType={'editor'}
    width={'100%'} />

## Props

### `defaultTextStyle`

`defaultTextStyle` is a convenience property. Whatever is passed will automatically be assigned to Pixi.js's [`TextStyle.defaultTextStyle`][pixi-defaultTextStyle].

> [!NOTE]
> This property **is not retroactive**. It will only apply to text components created after `defaultTextStyle` is set. Any text components created before setting `defaultTextStyle` will retain the base styles they had before `defaultTextStyle` was changed.

### `extensions`

`extensions` is an array of extensions to be loaded. Adding and removing items from this array will automatically load/unload the extensions. The first time this is handled happens before the application is initialised. See Pixi.js's [`extensions`][pixi-extensions] documentation for more info on extensions.

### `resizeTo`

The `<Application>` component supports the `resizeTo` property, with some additional functionality: it can accept any HTML element **or** it can take a React `ref` directly.

<EmbeddedEditor
    files={{ 'App.js': ExampleAppResizeTo }}
    viewType={'editor'}
    width={'100%'} />

[pixi-Application]: https://pixijs.download/release/docs/app.Application.html
[pixi-ApplicationOptions]: https://pixijs.download/release/docs/app.ApplicationOptions.html
[pixi-defaultTextStyle]: https://pixijs.download/release/docs/text.TextStyle.html#defaultTextStyle
[pixi-extensions]: https://pixijs.download/release/docs/extensions.html
