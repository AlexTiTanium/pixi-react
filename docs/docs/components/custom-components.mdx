---
sidebar_position: 3
title: Custom Components
---

import { EmbeddedEditor } from '@site/src/components/Editor/EmbeddedEditor';
import ExampleApp from '!!raw-loader!../../src/examples/custom-components/App';

Custom components are supported via the `extend` API. For example, you can create a `<viewport>` component using the [`pixi-viewport`][pixi-viewport] library:

<EmbeddedEditor
    files={{ 'App.js': ExampleApp }}
    viewType={'editor'}
    width={'100%'} />

The `extend` API will teach `@pixi/react` about your components, but TypeScript won't know about them nor their props. If you're using Typescript, check out our [docs for Typescript Users][typescript].

[pixi-viewport]: https://github.com/pixijs-userland/pixi-viewport
[typescript]: /typescript
