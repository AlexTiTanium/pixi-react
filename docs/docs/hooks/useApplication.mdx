import { EmbeddedEditor } from '@site/src/components/Editor/EmbeddedEditor';
import ExampleUseApplicationAccessible from '!!raw-loader!../../src/examples/useApplication-accessible/App';
import ExampleUseApplicationInaccessible from '!!raw-loader!../../src/examples/useApplication-inaccessible/App';

`useApplication` allows access to the parent `PIXI.Application` created by the `<Application>` component. This hook _will not work_ outside of an `<Application>` component. Additionally, the parent application is passed via [React Context][react-useContext]. This means `useApplication` will only work appropriately in _child components_, and in the same component that creates the `<Application>`.

For example, the following example `useApplication` **will not** be able to access the parent application:

<EmbeddedEditor
    files={{ 'App.js': ExampleUseApplicationInaccessible }}
    viewType={'editor'}
    width={'100%'} />

Here's a working example where `useApplication` **will** be able to access the parent application:

<EmbeddedEditor
    files={{ 'App.js': ExampleUseApplicationAccessible }}
    viewType={'editor'}
    width={'100%'} />

[react-useContext]: https://react.dev/reference/react/useContext
