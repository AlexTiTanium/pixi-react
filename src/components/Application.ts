import {
    type Application as PixiApplication,
    extensions as PixiExtensions,
    TextStyle,
} from 'pixi.js';
import {
    createElement,
    forwardRef,
    type ForwardRefRenderFunction,
    type MutableRefObject,
    useCallback,
    useEffect,
    useRef,
} from 'react';
import { createRoot } from '../core/createRoot';
import { roots } from '../core/roots';
import { processUnmountQueue } from '../helpers/processUnmountQueue';
import { queueForUnmount } from '../helpers/queueForUnmount';
import { unqueueForUnmount } from '../helpers/unqueueForUnmount';
import { useIsomorphicLayoutEffect } from '../hooks/useIsomorphicLayoutEffect';
import { type ApplicationProps } from '../typedefs/ApplicationProps';

const originalDefaultTextStyle = { ...TextStyle.defaultTextStyle };

/** Creates a React root and renders a Pixi application. */
export const ApplicationFunction: ForwardRefRenderFunction<PixiApplication, ApplicationProps> = (
    props,
    forwardedRef,
) =>
{
    const {
        attachToDevTools,
        children,
        className,
        defaultTextStyle,
        extensions,
        onInit,
        resizeTo,
        ...applicationProps
    } = props;

    const applicationRef: MutableRefObject<PixiApplication | null> = useRef(null);
    const canvasRef: MutableRefObject<HTMLCanvasElement | null> = useRef(null);
    const extensionsRef: MutableRefObject<Set<any>> = useRef(new Set());

    const updateResizeTo = useCallback(() =>
    {
        const application = applicationRef.current;

        if (application)
        {
            if (resizeTo)
            {
                if ('current' in resizeTo)
                {
                    if (resizeTo.current instanceof HTMLElement)
                    {
                        application.resizeTo = resizeTo.current;
                    }
                }
                else
                {
                    application.resizeTo = resizeTo;
                }
            }
            else
            {
                // @ts-expect-error Actually `resizeTo` is optional, the types are just wrong. 🤷🏻‍♂️
                application.resizeTo = undefined;
            }
        }
    }, [resizeTo]);

    const handleInit = useCallback((application: PixiApplication) =>
    {
        processUnmountQueue();

        if (forwardedRef && ('current' in forwardedRef))
        {
            forwardedRef.current = application;
        }

        applicationRef.current = application;
        updateResizeTo();
        onInit?.(application);

        if (attachToDevTools)
        {
            const globalScope = globalThis as any;

            globalScope.__PIXI_APP__ = application;

            import('pixi.js').then((pixi) =>
            {
                globalScope.__PIXI_DEVTOOLS__ = {
                    app: application,
                    pixi,
                    renderer: application.renderer,
                    stage: application.stage,
                };
            });
        }
    }, [onInit]);

    useIsomorphicLayoutEffect(() =>
    {
        if (extensions)
        {
            const extensionsToHandle = [...extensions];
            const extensionsState = extensionsRef.current;

            // Check for extensions that have been removed from the array
            for (const extension of extensionsState.values())
            {
                const extensionIndex = extensionsToHandle.indexOf(extension);

                // If the extension is no longer in the array, we'll remove it from Pixi.js
                if (extensionIndex === -1)
                {
                    PixiExtensions.remove(extension);
                    extensionsState.delete(extension);
                }

                // Since the extension already existed in the state, we can remove it to prevent any further handling
                extensionsToHandle.splice(extensionIndex, 1);
            }

            // Load any remaining extensions.
            for (const extension in extensionsToHandle)
            {
                PixiExtensions.add(extension);
                extensionsState.add(extension);
            }
        }
    }, [extensions]);

    useIsomorphicLayoutEffect(() =>
    {
        const canvasElement = canvasRef.current;

        if (canvasElement)
        {
            let root = roots.get(canvasElement);

            if (!root)
            {
                root = createRoot(canvasElement, {}, handleInit);
            }

            root.render(children, applicationProps);
        }
    }, [
        applicationProps,
        children,
        handleInit,
        resizeTo,
    ]);

    useIsomorphicLayoutEffect(() =>
    {
        updateResizeTo();
    }, [resizeTo]);

    useIsomorphicLayoutEffect(() =>
    {
        if (defaultTextStyle)
        {
            Object.assign(TextStyle.defaultTextStyle, defaultTextStyle);
        }
        else
        {
            Object.assign(TextStyle.defaultTextStyle, originalDefaultTextStyle);
        }
    }, [defaultTextStyle]);

    // eslint-disable-next-line consistent-return
    useEffect(() =>
    {
        const canvasElement = canvasRef.current;

        if (canvasElement)
        {
            unqueueForUnmount(canvasElement);

            return () =>
            {
                queueForUnmount(canvasElement);
            };
        }
    }, []);

    return createElement('canvas', {
        className,
        ref: canvasRef,
    });
};

ApplicationFunction.displayName = 'Application';

export const Application = forwardRef(ApplicationFunction);
