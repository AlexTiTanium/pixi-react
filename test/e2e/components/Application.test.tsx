import { type Application as PixiApplication } from 'pixi.js';
import { useEffect } from 'react';
import {
    createContext,
    createRef,
    useContext,
    useEffect,
} from 'react';
import {
    describe,
    expect,
    it,
    vi,
} from 'vitest';
import { Application } from '../../../src/components/Application';
import { roots } from '../../../src/core/roots';
import { useApplication } from '../../../src/hooks/useApplication';
import { type ApplicationRef } from '../../../src/typedefs/ApplicationRef';
import { isAppMounted } from '../../utils/isAppMounted';
import { render } from '@testing-library/react';

describe('Application', () =>
{
    it('mounts correctly', async () =>
    {
        const renderer = await act(async () => render(<Application />));

        expect(renderer.container).toMatchSnapshot();
    });

    it('forwards its ref', async () =>
    {
        const onInitSpy = vi.fn();
        const ref = createRef<ApplicationRef>();

        await act(async () => render((
            <Application
                ref={ref}
                onInit={onInitSpy} />
        )));

        await expect.poll(() => onInitSpy.mock.calls.length).toEqual(1);

        expect(ref.current?.getApplication()).toBeInstanceOf(PixiApplication);
        expect(ref.current?.getCanvas()).toBeInstanceOf(HTMLCanvasElement);
    });

    it('forwards context', async () =>
    {
        const onInitSpy = vi.fn();
        const ParentContext = createContext<boolean>(null!);
        let receivedValue!: boolean;

        function Test()
        {
            receivedValue = useContext(ParentContext);

            return null;
        }

        await act(async () => render((
            <ParentContext.Provider value={true}>
                <Application onInit={onInitSpy}>
                    <Test />
                </Application>
            </ParentContext.Provider>
        )));

        await expect.poll(() => onInitSpy.mock.calls.length).toEqual(1);

        expect(receivedValue).toBe(true);
    });

    describe('onInit', () =>
    {
        it('runs the callback once', async () =>
        {
            const onInitSpy = vi.fn();

            const TestComponent = () => (
                <Application onInit={onInitSpy} />
            );

            render(<TestComponent />);

            await expect.poll(() => onInitSpy.mock.calls.length).toEqual(1);
        });
    });

    describe('unmount', () =>
    {
        it('unmounts after init', async () =>
        {
            let testApp = null as any as PixiApplication;
            let testAppIsInitialised = false;

            const TestChildComponent = () =>
            {
                const {
                    app,
                    isInitialised,
                } = useApplication();

                useEffect(() =>
                {
                    testApp = app;
                    testAppIsInitialised = isInitialised;

                    return () =>
                    {
                        testApp = app;
                        testAppIsInitialised = isInitialised;
                    };
                }, [
                    app,
                    isInitialised,
                ]);

                return null;
            };

            const TestComponent = () => (
                <Application>
                    <TestChildComponent />
                </Application>
            );

            expect(roots.size).toEqual(0);

            const { unmount } = render(<TestComponent />);

            expect(roots.size).toEqual(1);

            await expect.poll(() => testAppIsInitialised).toEqual(true);

            unmount();

            expect(roots.size).toEqual(0);

            await expect.poll(() => isAppMounted(testApp)).toBeFalsy();
        });

        it('unmounts during init', async () =>
        {
            let testApp = null as any as PixiApplication;
            let testAppIsInitialised = false;

            const TestChildComponent = () =>
            {
                const {
                    app,
                    isInitialised,
                } = useApplication();

                useEffect(() =>
                {
                    testApp = app;
                    testAppIsInitialised = isInitialised;

                    return () =>
                    {
                        testApp = app;
                        testAppIsInitialised = isInitialised;
                    };
                }, [
                    app,
                    isInitialised,
                ]);

                return null;
            };

            const TestComponent = () => (
                <Application>
                    <TestChildComponent />
                </Application>
            );

            expect(roots.size).toEqual(0);

            const { unmount } = render(<TestComponent />);

            expect(roots.size).toEqual(1);

            expect(testAppIsInitialised).toBeFalsy();

            unmount();

            await expect.poll(() => isAppMounted(testApp)).toBeFalsy();

            expect(roots.size).toEqual(0);
        });
    });
});
