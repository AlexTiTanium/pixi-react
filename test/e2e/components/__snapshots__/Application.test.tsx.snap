// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Application > mounts correctly 1`] = `
<div>
  <canvas />
</div>
`;
